crates/marketplace/src/service.rs

```diff
@@ -337,22 +134,171 @@ where
 
         Ok(app)
     }
+
+    /// Spawns an event loop handling HotShot events from the provided stream.
+    /// Returns a handle for the spawned task.
+    pub fn start_event_loop(
+        &self,
+        event_stream: impl Stream<Item = Event<Types>> + Unpin + Send + 'static,
+    ) -> JoinHandle<anyhow::Result<()>> {
+        async_compatibility_layer::art::async_spawn(Self::event_loop(
+            self.coordinator.clone(),
+            self.hooks.clone(),
+            event_stream,
+        ))
+    }
+
+    /// Internal implementation of the event loop, drives the underlying coordinator
+    /// and runs hooks
+    async fn event_loop(
+        coordinator: Arc<BuilderStateCoordinator<Types>>,
+        hooks: Arc<Hooks>,
+        mut event_stream: impl Stream<Item = Event<Types>> + Unpin + Send + 'static,
+    ) -> anyhow::Result<()> {
+        loop {
+            let Some(event) = event_stream.next().await else {
+                anyhow::bail!("Event stream ended");
+            };
+
+            hooks.handle_hotshot_event(&event).await;
+
+            match event.event {
+                EventType::Error { error } => {
+                    tracing::error!("Error event in HotShot: {:?}", error);
+                }
+                EventType::Transactions { transactions } => {
+                    let transactions = hooks.process_transactions(transactions).await;
+
+                    // TODO: record results
```

Comment:
What do we want to record here?

Reply:
I remember I was thinking about the transaction status API, no idea why I'd add the comment though, as it has no relation to this PR 😅 . Removed it.